cellar_rebalancer/src/allocation.rs

```diff
@@ -0,0 +1,157 @@
+    proto::AllocationPrecommit {
+        hash: hasher.hash,
+        cellar_id: pair_id,
+    }
+}
+
+pub async fn cellar_query_client() -> Connections {
+    let config = APP.config();
+    let timeout = Duration::from_secs(10);
+    let try_base = AllocationQueryClient::connect(config.cosmos.grpc.clone()).await;
+    let (grpc, contact) = match try_base {
+        Ok(val) => (
+            Some(val),
+            Some(Contact::new(&config.cosmos.grpc, timeout, &config.cosmos.prefix).unwrap()),
+        ),
+        Err(e) => {
+            warn!(
+                "Failed to access Cosmos gRPC with {:?} and create connections",
+                e
+            );
+            (None, None)
+        }
+    };
+    Connections { grpc, contact }
+}
+
+pub async fn decide_rebalance(
+    tick_range: Vec<proto::TickRange>,
+    pair_id: U256,
+    eth_gas_price: u64,
+) {
+    if std::env::var("CELLAR_DRY_RUN").expect("Expect CELLAR_DRY_RUN var") == "TRUE" {
+        ()
+    } else {
+        let connections = cellar_query_client().await;
+        let contact = connections.contact.unwrap();
+        let grpc_client = &mut connections.grpc.unwrap();
+
+        let config = APP.config();
+        let cosmos_gas_price = config.cosmos.gas_price.as_tuple();
+        let allocation = to_allocation(tick_range, pair_id.to_string(), eth_gas_price);
+
+        let name = &config.keys.rebalancer_key;
+        let cosmos_key = config.load_deep_space_key(name.clone());
+        let delegate_cosmos_address = cosmos_key.to_address(&contact.get_prefix()).unwrap();
+
+        let fee = Coin {
+            denom: cosmos_gas_price.1,
+            amount: 0u32.into(),
+        };
+
+        match timeout(Duration::from_secs(100), async {
+            loop {
+                // Waiting for new vote period
+                let res = somm_send::query_commit_period(grpc_client).await;
+                if let Ok(val) = res {
+                    if val.vote_period_start == val.current_height {
+                        break val;
+                    }
+                }
+                sleep(Duration::from_secs(1)).await;
+            }
+        })
+        .await
+        {
+            Ok(val) => {
+                // Sending Pre-commits
+                somm_send::send_precommit(
+                    &contact,
+                    delegate_cosmos_address,
+                    cosmos_key,
+                    fee.clone(),
+                    vec![allocation_precommit(&cosmos_key, &allocation, pair_id.to_string()).await],
+                )
+                .await
+                .unwrap();
+            }
+            Err(e) => {
+                println!("Couldn't Send Precommit {:?}", e);
+            }
+        }
+
+        match timeout(Duration::from_secs(100), async {
+            loop {
+                // Checking Pre-commits for validators
+                let res = somm_send::query_allocation_precommits(grpc_client).await;
+                if let Ok(val) = res {
```

Comment:
Let's add a one-block delay here before proceeding to `break`, so we submit our commit in the block after our precommit.